### PR TITLE
libfaiss: Update to version 1.6.4, add Python ports

### PR DESCRIPTION
**Pull request by essandess**
libfaiss: Update to version 1.6.4, add Python ports

* Update to version 1.6.4
* Add python ports

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.x
Xcode 8.x

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: math/libfaiss/Portfile [hunk 1]
```diff
@@ -44,19 +46,127 @@ build.env-append    "SHAREDFLAGS=-Wl,-F. -bundle -undefined dynamic_lookup"
...
+    variant openblas description {Use OpenBLAS instead of Apple's Accelerate framework} {
+        depends_lib-append  port:OpenBLAS
+        build.env-delete    "BLASLDFLAGS=-framework Accelerate"
+        build.env-append    BLASLDFLAGS=${prefix}/lib/libopenblas.dylib
+    }
+
+    default_variants +openblas
+}
+
+# Python subports
```
Comment (reneeotten) [inline math/libfaiss/Portfile:62]: perhaps the example Portfile I suggested earlier today wasn't the best example, sorry.... What you've added here now is not very readable, which is understandable if the port actually needs all these variables and stuff.

I would suggest to do it a little different in this case: first just define the subports and then use in the ` if {[string match "py*" ${subport}]} {` block the `python` PG to make things easier. If you do that you can re-use all the variables like ${python.bin} and such from the PG. Perhaps something like this (of course it still needs the configure/build parts):
```
# create Python subports
foreach v {37 38} {
    subport py${v}-${github.project} {}
}

# Python bindings for supported Python versions
if {[string match "py*" ${subport}]} {
    PortGroup           python 1.0

    categories-append   math
    description         Python ${python.branch} bindings for ${description}
    long_description    This package provides Python ${python.branch} bindings, \
                        modules, and scripts.

    python.default_version ${python.version}

    depends_build-append \
                        port:py${python.version}-setuptools \
                        port:swig-python
    depends_lib-append  port:${name} \
                        port:py${python.version}-numpy
   
    <whatever else is needed>
}

Comment (essandess) [reply math/libfaiss/Portfile]: Thanks done. I had to move `python.default_version` into the `support` to avoid the default version `38` when `python.versions` is undefined.

Comment (reneeotten) [reply math/libfaiss/Portfile]: yes, good catch!

File: math/libfaiss/Portfile [hunk 2]
```diff
@@ -24,39 +22,134 @@ long_description    Library for efficient similarity search and clustering of \
...
+    configure.env-append \
+                        PYTHON=${python.bin}
+
+    configure.args-append \
+                        --with-python=${python.bin} \
+                        --with-swig=${prefix}/bin/swig
+
+    pre-build {
+        system -W ${worksrcpath} ${build.cmd}
+    }
```
Comment (reneeotten) [inline math/libfaiss/Portfile:105]: this now builds the library again - we don't want to do that as you already added a dependency on `libfaiss` so it should use the library that was already build by that port instead of building it again

File: math/libfaiss/Portfile [hunk 3]
```diff
@@ -24,39 +22,134 @@ long_description    Library for efficient similarity search and clustering of \
...
+                        port:swig-python
+    depends_lib-append  port:${name} \
+                        port:py${python.version}-numpy
+
+    post-extract {
+        # fix destroot commands like `$(PYTHON) setup.py install`
+        reinplace -E \
+            "s|^(\[\[:space:]]+\\\$\\(PYTHON\\)\[\[:space:]]+setup\.py\[\[:space:]]+)(install)|\\1 --no-user-cfg \\2 ${destroot.destdir}|" \
+            ${worksrcpath}/python/Makefile
+    }
```
Comment (reneeotten) [inline math/libfaiss/Portfile:92]: I'd prefer this as a patch instead of an (unreadable) `reinplace`

File: math/libfaiss/Portfile [hunk 4]
```diff
@@ -24,39 +22,134 @@ long_description    Library for efficient similarity search and clustering of \
...
 compilers.setup
 
-use_configure       no
+depends_lib-append  port:libomp
+
+configure.args      --without-cuda
+
+build.env-append    CC=${configure.cxx} \
+                    "CFLAGS=-fPIC -m64 -Wall -g -O3 -msse4 -mpopcnt -fopenmp \
+                        -Wno-sign-compare -std=c++11 -I/usr/include/malloc/" \
```
Comment (reneeotten) [inline math/libfaiss/Portfile:42]: you're adding `-std=c++11` here so you likely want to make sure you get a compatible compiler (i.e., use `compiler.cxx_standard   2011`)

File: math/libfaiss/Portfile [hunk 5]
```diff
@@ -24,39 +22,134 @@ long_description    Library for efficient similarity search and clustering of \
...
 
+checksums           rmd160  3b1180dc7ceadc484bb86cfe32fb5cd1c4a694b9 \
+                    sha256  6962641a9183e12d394aa2f9c443593734a373d60c61fd0633edda5029d2fa78 \
+                    size    2762193
+
 compilers.choose    cxx
 # The Apple native llvm compiler does not support OpenMP.
 compiler.blacklist  llvm-gcc-4.2
-compiler.fallback   macports-gcc-7
+compiler.fallback   macports-gcc-9
```
Comment (reneeotten) [inline math/libfaiss/Portfile:32]: this doesn't seem right... for example if I do `port info libfaiss +gcc10` I both get `gcc9` and `gcc10` as build dependencies. I am not completely sure how this is supposed to work and why it's needed, but I don't think it does what you want right now

File: math/libfaiss/Portfile (hunk 2 above)
Comment (essandess) [reply math/libfaiss/Portfile:105]: The make system requires this: https://github.com/facebookresearch/faiss/blob/master/INSTALL.md

File: math/libfaiss/Portfile (hunk 5 above)
Comment (essandess) [reply math/libfaiss/Portfile:32]: I don’t know either. Should these lines be deleted?

File: math/libfaiss/Portfile (hunk 3 above)
Comment (essandess) [reply math/libfaiss/Portfile:92]: Patches are a PITA to create and maintain and this works.

File: math/libfaiss/Portfile [hunk 6]
```diff
@@ -24,39 +22,134 @@ long_description    Library for efficient similarity search and clustering of \
...
+                                \"test_meta_index.py\", \
+                                \"test_omp_threads_py.py\", \
+                                \"test_oom_exception.py\", \
+                                \"test_referenced_objects.py\", \
+                                \"test_standalone_codec.py\"]; \
+                            \[ printout(scpt) for scpt in test_scripts ];'"
+    test.target
+    test.env-append \
+                    "PATH=$env(PATH):${python.prefix}/bin" \
+                    PYTHONPATH=${worksrcpath}/build/lib
```
Comment (reneeotten) [inline math/libfaiss/Portfile:152]: the tests fail with `ModuleNotFoundError: No module named 'faiss'` - in the PR checklist it states that the tests work (it might if you have the port already installed though). It should likely be `PYTHONPATH=${worksrcpath}/build/lib` instead.

File: math/libfaiss/Portfile (hunk 3 above)
Comment (reneeotten) [reply math/libfaiss/Portfile:92]: you know also that it's the preferred way of doing this in MacPorts.... besides that you should patch that file anyway to make sure it includes the library that was build already instead doing this again (see my comment above as well)

File: math/libfaiss/Portfile (hunk 2 above)
Comment (essandess) [reply math/libfaiss/Portfile]: I just added a symlink to `libfaiss.a` in `worksdrcpath` and resolved all these issues.

File: math/libfaiss/Portfile (hunk 3 above)
Comment (essandess) [reply math/libfaiss/Portfile:92]: See above. A simple symlink works.

File: math/libfaiss/Portfile (hunk 5 above)
Comment (essandess) [reply math/libfaiss/Portfile:32]: The blacklist can be deleted. The fallback is necessary for the build. Not sure why.

File: math/libfaiss/Portfile (hunk 6 above)
Comment (essandess) [reply math/libfaiss/Portfile]: Fixed thanks.

File: math/libfaiss/Portfile (hunk 4 above)
Comment (essandess) [reply math/libfaiss/Portfile:41]: Added thanks.

File: math/libfaiss/Portfile [hunk 7]
```diff
@@ -1,14 +1,12 @@
...
 
 PortSystem          1.0
-PortGroup           github 1.0
 PortGroup           compilers 1.0
+PortGroup           github 1.0
 
-github.setup        facebookresearch faiss 56383610bcb982d6591e2e2bea3516cb7723e04a
+github.setup        facebookresearch faiss 1.5.3 v
 name                libfaiss
-version             20180214
```
Comment (reneeotten) [inline math/libfaiss/Portfile:9]: I think there is an issue here with the versioning in the sense that `20180214` will be considered "newer" than `1.5.3.`. Please check and install the current version in `master` first, then change to your updated-libfaiss-branch, do a `portindex` and `port outdated`. If the `libfaiss` port doesn't show up there, this is indeed an issue and you should add `epoch 1`.

Comment (essandess) [reply math/libfaiss/Portfile:9]: The current `libfaiss` cannot be installed on current macOS because of a build error.

IS there a way of checking this issue without an install?

Comment (essandess) [reply math/libfaiss/Portfile:9]: I just went ahead and added `epoch 1`.

File: math/libfaiss/Portfile (hunk 3 above)
Comment (essandess) [reply math/libfaiss/Portfile:93]: I looked into a patch but that's a very complicated solution here because the patch involves `${prefix}`. I simple `reinplace` is much easier and transparent.
```
        $(PYTHON) setup.py  --no-user-cfg install --prefix=/opt/local/Library/Frameworks/Python.framework/Versions/3.7 --root=/opt/local/var/macports/build/_opt_local_ports_python_py-faiss/py37-faiss/work/destroot
```

Comment (kencu) [reply math/libfaiss/Portfile:93]: The usual thing we do to get around the prefix issue is to patch in a placeholder like @@PREFIX@@ and reinplace the placeholder post-patch. At least we can read that.

Few can tell what this is doing (I don't have the slightest idea what it's up to), it's difficult to debug it when it breaks (if you know it's broken), and the last guy who tried this made some serious errors doing it that were not clear until I ran the thing and generated a diff to see the wreckage.

Comment (essandess) [reply math/libfaiss/Portfile:99]: *sigh* no appreciation for BSD sed-specific regex syntax wrapped in tcl…

Changed it to a patch, thanks.

**Comment by essandess** [comment]
@kencu To the best of my knowledge this is robust and ready to merge. Would you please take a look? There are other PRs that have this as a dependency.

**Comment by kencu** [comment]
If I commit this, I will never hear the end of it from @ryandesign . 

Very sorry, you'll have to wait for him to take up the various issues with libstdc++, seemingly only a static library, forcing the build of a c++ program with g++-10.

Now that I've said this, he'll likely just commit it and ask why I was unnecessarily worried about it.

**Comment by essandess** [comment]
> Very sorry, you'll have to wait for @ryandesign to take up the various issues with libstdc++, seemingly only a static library, forcing the build of a c++ program with g++-10, etc.

FWIW these were all in the original Portfile from the faiss team. Perhaps they’re unnecessary with the latest cmake build. What are the preferred defaults?

**Comment by kencu** [comment]
For a port that provides a c++ library, the port should 

- build with clang++ 
- the  c++ standard library should be libc++
- should build a dynamic library so that updates flow through to all the dependencies automatically.

There might be deviations from this allowed for certain unique situations, but each would have to be justified and an explanation for it put in the portfile.

File: math/libfaiss/Portfile (hunk 7 above)
Comment (ryandesign) [reply math/libfaiss/Portfile:9]: > The current `libfaiss` cannot be installed on current macOS because of a build error.

Unless you are saying that no previous version of this port could ever be installed on any system—which I can see by looking at https://packages.macports.org/libfaiss/ is not the case—you should increase the epoch.

> I just went ahead and added `epoch 1`.

You added `epoch 0`, which is the default; to have any effect it would need to be `epoch 1`.

File: math/libfaiss/Portfile [hunk 8]
```diff
@@ -60,3 +62,92 @@ variant openblas description {Use OpenBLAS instead of Apple's Accelerate framewo
...
+                    -DPython3_EXECUTABLE=${python.bin} \
+                    -DSWIG_EXECUTABLE=${prefix}/bin/swig
+    configure.pre_args-delete \
+                    --prefix=${prefix}
+    configure.pre_args-replace \
+                    {-DCMAKE_C_COMPILER="$CC"} \
+                    -DCMAKE_C_COMPILER=${configure.cc}
+    configure.pre_args-replace \
+                    {-DCMAKE_CXX_COMPILER="$CXX"} \
+                    -DCMAKE_CXX_COMPILER=${configure.cxx}
```
Comment (ryandesign) [inline math/libfaiss/Portfile:116]: Why is this needed? The existing lines are supposed to set the CMake C and C++ compilers to whatever the CC and CXX environment variables are. Why is that not sufficient?

**Comment by ryandesign** [comment]
> What are the preferred defaults?

The preferred defaults are set by MacPorts. You get them by not adding lines to the portfile to override things.

This port is called "libfaiss". I assume that means it installs a library that is meant to be used by others. As such, if the library use C++, it is important that it uses the same C++ library that will be used by whoever is trying to use it, because C++ libraries cannot be mixed and matched. The default C++ standard library on 10.9 and later (and on 10.6-10.8 in MacPorts) is libc++, not libstdc++.

There was a comment in the port, that you removed, that said:

> The Apple native llvm compiler does not support OpenMP

Perhaps that is why the port does all the strange things it does. After this port was added, MacPorts gained some support for knowing which compilers support openmp. Maybe setting `compiler.openmp_version` to the correct value would allow MacPorts to select the right compiler for you and remove the need for the compilers portgroup and all of the other compiler related code from the portfile.

**Comment by essandess** [comment]
Thanks for all the guidance. IIII've pushed a mod that addresses your comments subject to these issues:

1. Turning off the `compiler` directives (specifically `compiler.fallback macports-gcc-10`) throws this error:
```
-- Detecting CXX compile features - done
CMake Error at /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
Call Stack (most recent call first):
  /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  /opt/local/share/cmake-3.18/Modules/FindOpenMP.cmake:529 (find_package_handle_standard_args)
  faiss/CMakeLists.txt:147 (find_package)
```

2. Just using
```
compiler.cxx_standard    2011
compiler.openmp_version 2.5
```
Creates binaries that throw this openmp linking error: `Symbol not found: ___kmpc_barrier`

3. Omitting the directive
```
configure.cxx_stdlib    libstdc++
```
creates a broken port with this error:
```
py37-faiss is using libstdc++ (this installation is configured to use libc++)
```

> For a port that provides a c++ library, the port should
> 
> * build with clang++
> * the  c++ standard library should be libc++
> * should build a dynamic library so that updates flow through to all the dependencies automatically.
> 
> There might be deviations from this allowed for certain unique situations, but each would have to be justified and an explanation for it put in the portfile.

File: math/libfaiss/Portfile (hunk 8 above)
Comment (essandess) [reply math/libfaiss/Portfile]: Thanks not needed.

File: math/libfaiss/Portfile (hunk 7 above)
Comment (essandess) [reply math/libfaiss/Portfile:9]: Thanks done.

**Review by kencu — COMMENTED** [review]
> Note these issues for posterity:
> 
> 1. Turning off the `compiler` directives (specifically `compiler.fallback macports-gcc-10`) throws this error:
> 
> ```
> -- Detecting CXX compile features - done
> CMake Error at /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
>   Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
> Call Stack (most recent call first):
>   /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
>   /opt/local/share/cmake-3.18/Modules/FindOpenMP.cmake:529 (find_package_handle_standard_args)
>   faiss/CMakeLists.txt:147 (find_package)
> ```
> 
> 1. Just using
> 
> ```
> compiler.cxx_standard    2011
> compiler.openmp_version 2.5
> ```
> 
> Creates binaries that throw this openmp linking error: `Symbol not found: ___kmpc_barrier`
> 
> 1. Omitting the directive
> 
> ```
> configure.cxx_stdlib    libstdc++
> ```
> 
> creates a broken port with this error:
> 
> ```
> py37-faiss is using libstdc++ (this installation is configured to use libc++)
> ```

Yes, this port is a bloody mess. All these things will have to be properly fixed. Good on you for taking it on. Ryan will help you out.

**Comment by essandess** [comment]
> @kencu Yes, this port is a bloody mess. All these things will have to be properly fixed. Good on you for taking it on. Ryan will help you out.

My main point is that I believe that all these issues have been resolved and the port is ready to merge.

I just added that stuff for posterity in case someone ever wants tackle migrating to clang with openmp in the future.

**Comment by kencu** [comment]
> > @kencu Yes, this port is a bloody mess. All these things will have to be properly fixed. Good on you for taking it on. Ryan will help you out.
> 
> My main point is that I believe that all these issues have been resolved and the port is ready to merge.
> 
> I just added that stuff for posterity in case someone ever wants tackle migrating to clang with openmp in the future.

Steve -- maybe Ryan will commit this, but these things are not resolved. You can't force libstdc++ just because it doesn't build unless you do...

It's a mess, needs to be fixed to build with clang, needs to build against libc++, and needs openmp to work properly (which many ports do not, as Apple's clang doesn't support OpenMP, so there is not much impetus out there in the vast open-source community to fix software to build with OpenMP on Apple systems just yet).

But we'll see if Ryan pushes this for you.

**Comment by essandess** [comment]
> > > @kencu Yes, this port is a bloody mess. All these things will have to be properly fixed. Good on you for taking it on. Ryan will help you out.
> > 
> > 
> You can't force libstdc++ just because it doesn't build unless you do...

`libfaiss` doesn’t use `libstdc++`, only thy Python module does, which breaks without it. But this should be fine because you don’t link against a Python module.

> It's a mess, needs to be fixed to build with clang, needs to build against libc++, and needs openmp to work properly (which many ports do not, as Apple's clang doesn't support OpenMP, so there is not much impetus out there in the vast open-source community to fix software to build with OpenMP on Apple systems just yet).

Upstream needs OpenMP, and this port is a dependency of several other projects. Whether clang or gcc is required to build it is irrelevant to its users and users of its dependents.

**Comment by kencu** [comment]
OK, if that's the last on it, there you sit. Up to Ryan.

**Comment by essandess** [comment]
@kencu @ryandesign Okay now I believe that all these issue are resolved after the last [force-push](https://github.com/macports/macports-ports/compare/44bd8c5a10d00bf4d4f101446cb9afcaa9348dc3..65026f9221725545d53f1b7979ab5561c0845e8f). Let's see if the builedbot builds…

**Comment by kencu** [comment]
> @kencu @ryandesign Okay now I believe that all these issue are resolved after the last [force-push](https://github.com/macports/macports-ports/compare/44bd8c5a10d00bf4d4f101446cb9afcaa9348dc3..65026f9221725545d53f1b7979ab5561c0845e8f). Let's see if the buildbot builds…

Whoa! Now you're talkin' !

Somebody will tell you that you don't need this any more:
```
depends_lib-append  port:libomp

configure.ldflags-append \
                    -L${prefix}/lib/libomp \
                    -lomp
```

because it's (supposed to be) automatically set by this:
```
compiler.openmp_version 2.5
```
through yet more undocumented PortGroup magic.

**Comment by essandess** [comment]
These `configure.ldflags` are necessary. Without them, `libomp.dylib` doesn't get linked and the Python modules throw the error `Symbol not found: ___kmpc_barrier`.

This looks like either a MacPorts bug or a conflict with the combination of the other port groups `cmake` and `python`.

Either way, I believe this is ready to merge and ask that we move ahead to its dependents.


Without `configure.ldflags` :
```
During handling of the above exception, another exception occurred:
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_oom_exception.py:7: in <module>
    import faiss
../build/faiss/python/faiss/__init__.py:16: in <module>
    from .loader import *
../build/faiss/python/faiss/loader.py:39: in <module>
    from .swigfaiss import *
../build/faiss/python/faiss/swigfaiss.py:13: in <module>
    from . import _swigfaiss
E   ImportError: dlopen(/opt/local/var/macports/build/_opt_local_ports_math_libfaiss/py37-faiss/work/build/faiss/python/faiss/_swigfaiss.so, 2): Symbol not found: ___kmpc_barrier
E     Referenced from: /opt/local/var/macports/build/_opt_local_ports_math_libfaiss/py37-faiss/work/build/faiss/python/faiss/_swigfaiss.so
E     Expected in: flat namespace
E    in /opt/local/var/macports/build/_opt_local_ports_math_libfaiss/py37-faiss/work/build/faiss/python/faiss/_swigfaiss.so
```

> > @kencu @ryandesign Okay now I believe that all these issue are resolved after the last [force-push](https://github.com/macports/macports-ports/compare/44bd8c5a10d00bf4d4f101446cb9afcaa9348dc3..65026f9221725545d53f1b7979ab5561c0845e8f). Let's see if the buildbot builds…
> 
> Whoa! Now you're talkin' !
> 
> Somebody will tell you that you don't need this any more:
> 
> ```
> depends_lib-append  port:libomp
> 
> configure.ldflags-append \
>                     -L${prefix}/lib/libomp \
>                     -lomp
> ```
> 
> because it's (supposed to be) automatically set by this:
> 
> ```
> compiler.openmp_version 2.5
> ```
> 
> through yet more undocumented PortGroup magic.

**Comment by essandess** [comment]
Ticket about `compiler.openmp_version`: https://trac.macports.org/ticket/61390#ticket

**Comment by kencu** [comment]
Personally I think we've made you jump through quite enough hoops.

However, as this port is still not "perfect" and as I don't care to get blowback on github from committing it, we will wait for @ryandesign to get around to it.

**Comment by ryandesign** [comment]
> because it's (supposed to be) automatically set by this:
> 
> ```
> compiler.openmp_version 2.5
> ```
> 
> through yet more undocumented PortGroup magic.

`compiler.openmp_version` is a new MacPorts base feature (not portgroup) and is documented or at least mentioned in https://trac.macports.org/wiki/CompilerSelection. Improvements to documentation are always welcome.

**Comment by kencu** [comment]
Can you help poor Steve here get this over the finish line?

**Comment by ryandesign** [comment]
I wasn't going to comment on the `-I/usr/include/malloc` flag since it was already in the port before this PR, but now that you've removed it, and since it seems to build, that means this closes https://trac.macports.org/ticket/57420 and that could be added to the commit message, or we can just close the ticket after this gets merged.

> However, as this port is still not "perfect" and as I don't care to get blowback on github from committing it, we will wait for @ryandesign to get around to it.

The feedback that I provide on pull requests and commits is not intended to be critical of the author, it's intended to improve the user experience by giving ports as consistent a behavior as possible, and to improve the maintainer experience by making our ports' source code as consistent as possible. There is often "a MacPorts way" to do something, and doing it that way has various advantages, so when a developer has done something different, I want to point out how we would usually do it instead. My hope is that this feedback improves everyone's knowledge of how to write good portfiles so that over time the need for such feedback hopefully decreases. I encourage other developers to read PRs and commits and provide feedback as well.

You don't need to wait on me for feedback on any PR. I may have time to add a comment to a PR but that does not mean that I will have time to review any subsequent discussion on that PR in a timely manner. Any developer with commit access can and should merge PRs that appear to be correct and conform to our usual coding style and methods and are approved by the port's maintainer or have no maintainer response for 3 days.

**Comment by ryandesign** [comment]
When I try to install py37-faiss or py38-faiss on my system, I get:
```
CMake Error at /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find Python (missing: Python_NumPy_INCLUDE_DIRS NumPy) (found
  version "2.7.16")
Call Stack (most recent call first):
  /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  /opt/local/share/cmake-3.18/Modules/FindPython.cmake:436 (find_package_handle_standard_args)
  faiss/python/CMakeLists.txt:78 (find_package)
```
I am not sure why the CI builds succeed.

**Comment by essandess** [comment]
FWIW I do not see this on my system. I have seen errors like this on other ports, and I believe it has something to do with the cmake system finding the wrong python executable. It can be fixed by setting some cmake argument. See e.g. https://github.com/macports/macports-ports/pull/7791#issuecomment-660649668.

Looking at the build, I see this, suggesting that I set the wrong flag for the python executable.
```
  Manually-specified variables were not used by the project:

    CMAKE_C_COMPILER
    CMAKE_POLICY_DEFAULT_CMP0025
    CMAKE_POLICY_DEFAULT_CMP0060
    Python3_EXECUTABLE
```

I'll dig into this and get back…

> When I try to install py37-faiss or py38-faiss on my system, I get:
> 
> ```
> CMake Error at /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
>   Could NOT find Python (missing: Python_NumPy_INCLUDE_DIRS NumPy) (found
>   version "2.7.16")
> Call Stack (most recent call first):
>   /opt/local/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
>   /opt/local/share/cmake-3.18/Modules/FindPython.cmake:436 (find_package_handle_standard_args)
>   faiss/python/CMakeLists.txt:78 (find_package)
> ```
> 
> I am not sure why the CI builds succeed.

**Comment by essandess** [comment]
Per https://github.com/facebookresearch/faiss/blob/master/INSTALL.md#general-compilation-instructions, the correct cmake flag is `-DPython_EXECUTABLE`. Thanks for pointing out the error @ryandesign. Fix pushed.

**Comment by ryandesign** [comment]
Thanks! That works for me.
